### ISSUE TITLE
PythonTeX over SSH

### ISSUE THREAD

**Issue by TheChymera** [gpoore/pythontex#77]
I am quite enthusiastic about PythonTeX, and try to set up most of the figures in my presentations and manuscripts with it. Sadly, however, as the data with which I work gets bigger and bigger, I can no longer perform the analysis on my laptop/desktop computers on which I usually compile my latex documents. I can think of two solutions:

* compile the document on the remote hosts that can perform my data analysis.
* set up some ssh pipe which lets PythonTeX run the analysis on my server, fetch the figures, and compile the document locally.

The problem I can see with the first approach is that I would need to constantly do a lot of stuff over SSH in the editing/file creation process. 

* I would need to edit my text files over SSH (works, but the shortcut keys for compilation do not)
* I would need to have the resulting PDF opened in some PDF viewer that can open it via SSH (I do not know o any such viewer)
* I would constantly need to scp static images which I get from the web and would like to display in my document

The second approach seems preferable to me, but I would not know how to make pythontex do that. Do you?

**Comment by gpoore** [comment]
Here's another possibility to consider, depending on the restrictions of your data. Perhaps you can have the remote machines save a processed form of the data that is suitable for plotting. Then you could do the actual plotting locally. Something like this should be possible unless the actual plotting/visualization itself is very computationally intensive. This would have the advantage that changing the plots wouldn't require redoing the analysis.

The tricky thing with this approach would be making sure that the local copy of the analysis code stayed in sync with the remote version.

**Comment by TheChymera** [comment]
I had actually thought of that, but, in my opinion, one of the strengths of pythontex is that it is "smart" enough to decide whether to re-run the analysis depending on what files have changed since the last run. By making some intermediary data analysis steps static, I lose this feature for all steps that occur before that in my pipeline. As you have correctly observed, if - for instance - I add another subject to my dataset or change some preprocessing parameters, I would have to make sure (separately from the features that pythontex has in place for this) that my update is correctly represented at the intermediary step.

Long story short, I think this makes a lot of pythontex' power go to waste.

For the time being I have decided to mount my remote filesystem with sshfs and try to do all my latex stuff on the remote host. Needless to say this has a number of drawbacks including pdf viewers not automatically refreshing the file, having to call the latex/pythntex/bibtex compilation from the command line, but I think this is a good trade-off for keeping my data representation clean and "live".

Is there any hope of pythontex performing remote analysis in the future, or am I the first to bring this issue up?

**Comment by gpoore** [comment]
You are the first one to bring this up.

I think this could be done, but I'm not sure about a time frame for doing it. Part of the reason for asking about intermediary data analysis is that it makes things much easier. Here's what I would envision as a possible solution.
* One PythonTeX session contains the code for the remote analysis. There might be a need to set an option `remote=true` somewhere in the LaTeX document, perhaps with some additional ssh info and info about the remote files that are generated. When PythonTeX is run (maybe with a `--remote` flag), the code for this session is saved to a file, the file is transferred to the remote location, and then the file is executed. Ideally, there would be a check for modified dependencies somehow, so that execution would only occur when needed. Afterward, any new or modified output (intermediary data) would be transferred back to the local machine.
* In a separate PythonTeX session, the intermediary data is plotted. This keeps things fast when remote work isn't needed.

Again, if this is actually possible, I'm not sure about a time frame. In the meantime, some sort of makefile approach might be possible to keep remote work synced with local work, if your current approach becomes too difficult and you need to move back to local work.

I am going to be doing a major overhaul of PythonTeX over the next couple months to add new features, and will keep this in mind during that process.

**Comment by gzahl** [comment]
Any news about this? I face the same challenge. I would like to use pythontex, but probably i should call some workflow system from pythontex, which actually manages the data analysis pipeline?

**Comment by gpoore** [comment]
@gzahl I'm still open to considering adding support for this to PythonTeX, but nothing will happen in the near future.

I've been working on writing lower-level libraries that will make the next major PythonTeX  overhaul possible, and that has turned into a larger project than I expected. It will probably be at least 3-6 months before I'm back to working on significant PythonTeX development. And even once that happens, most of the initial work will be related to the new libraries.

**Comment by gzahl** [comment]
Okay, thanks for the ETA

**Comment by TheChymera** [comment]
For the time being, I have fallen back to the proposed approach of separtely running the heavy processing on the server, and saving intermediary data analysis outputs for download. This is however quite frustrating, as it makes my workflow considerably less streamlined, more difficult to document all-in-one-place, and less reproducible. I am looking forward to updates on this as well!
